### ISSUE TITLE
Pytest requires undocumented `Output.future` method to wait for tasks

### ISSUE THREAD

**Issue by H0R5E** [pulumi/pulumi#11994]
### What happened?

I was following the [Unit Testing](https://www.pulumi.com/docs/guides/testing/unit/) guide for Python but using pytest instead of unittest. I found that pytest would not wait for the result of the `apply` method on an output, unless I did an explicit `await` on the result of the `future` method of the returned `Output` object.

### Expected Behavior

I was expecting pytest to work in a similar manner to the example using the builtin unittest framework. 

### Steps to reproduce

I have a very simple Python function to build an S3 bucket, which is included in a package:

```python
from pulumi_aws import s3

def main():
    bucket = s3.Bucket('my-bucket')
    return bucket
```

I then wrote an equivalent test to the guide, but for pytest, as so:

```python
import pulumi

class MyMocks(pulumi.runtime.Mocks):
    def new_resource(self, args: pulumi.runtime.MockResourceArgs):
        return [args.name + '_id', args.inputs]
    def call(self, args: pulumi.runtime.MockCallArgs):
        return {}

pulumi.runtime.set_mocks(
    MyMocks(),
    preview=False,  # Sets the flag `dry_run`, which is true at runtime during a preview.
)

from bluebox_pulumi import main

@pulumi.runtime.test
def test_main():
    
    def check_tags(tags):
        assert tags is not None
        assert 'Name' in tags
    
    bucket = main()
    bucket.tags.apply(check_tags)
```

When you call pytest on the above, it incorrectly passes. To make pytest wait for the asynchronous calls to complete, I had to await the result of the future method, as so:

```python
@pulumi.runtime.test
async def test_main():
    
    def check_tags(tags):
        assert tags is not None
        assert 'Name' in tags
    
    bucket = main()
    await bucket.tags.apply(check_tags).future()
```
 
The result of the pytest run is then correctly given as `FAILED tests/test_main.py::test_main - assert None is not None`.

I had to dig into the code of `output.py` to find the `future` method and it's marked as being private ("Private implementation details - do not document") so I guess I'm not supposed to be using it, but I don't see another way of making pytest wait.

### Output of `pulumi about`

```
> pulumi about
CLI
Go Version   go1.19.5
Go Compiler  gc

Plugins
NAME    VERSION
aws     5.28.0
python  unknown

Host
OS       Microsoft Windows 10 Pro
Version  10.0.19044 Build 19044
Arch     x86_64

This project is written in python: executable='C:\Anaconda3\envs\_bluebox_emulator\python.exe' version='3.9.15
'

Backend
Name           pulumi.com
URL            https://app.pulumi.com/H0R5E
User           H0R5E
Organizations  H0R5E, dataonlygreater

Pulumi locates its logs in C:\Users\Work\AppData\Local\Temp by default
warning: Failed to get information about the Pulumi program's dependencies: exit status 2
warning: Failed to get information about the current stack: No current stack
warning: A new version of Pulumi is available. To upgrade from version '3.52.1' to '3.53.1', visit https://pulumi.com/docs/reference/install/ for manual instructions and release notes.
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Frassle** [comment]
We should see if we can integrate with pytest as we do for unittest. Or at least make the workaround public and documented (for reference, I don't think it's safe to assume that workaround _will_ be the current future() method)

**Comment by lazcamus** [comment]
The `@pulumi.runtime.test` decorator does the awaiting for you **if** you `return` the `Output`:

```python
@pulumi.runtime.test
async def test_main():
    
    def check_tags(tags):
        assert tags is not None
        assert 'Name' in tags
    
    bucket = main()
    await bucket.tags.apply(check_tags).future()
```

becomes
```python
@pulumi.runtime.test
async def test_main():
    
    def check_tags(tags):
        assert tags is not None
        assert 'Name' in tags
    
    bucket = main()
    return bucket.tags.apply(check_tags)
```

**Comment by julienp** [comment]
I tried reproducing this, and it looks like this was fixed in 3.112.0 with https://github.com/pulumi/pulumi/pull/15744, where we now properly await all outputs.

In 3.111.0 both pytest and unittest wrongly fail unless we return the output from the test. With 3.112.0, the test correctly fails even without the return.
